File: day-1/puzzle-1/solution.exs
{list1, list2} =
  File.read!("input.txt")
  |> String.split("\n")
  |> Enum.reduce({[], []}, fn line, {list1, list2} ->
    Regex.named_captures(~r/(?<left>\d+)\s+(?<right>\d+)/, line)
    |> case do
      %{"left" => left, "right" => right} ->
        {[String.to_integer(left) | list1], [String.to_integer(right) | list2]}

      _ ->
        {list1, list2}
    end
  end)

# Sort the list
list1 = Enum.sort(list1)
list2 = Enum.sort(list2)

# Pair them
pairs = Enum.zip([list1, list2])

total_distance =
  Enum.reduce(pairs, 0, fn {left, right}, total_distance ->
    total_distance + abs(right - left)
  end)

IO.puts("total distance = #{total_distance}")

File: day-1/puzzle-1/js/solution.mjs
import fs from 'node:fs';
// Read from file into variable
const fsPromises = fs.promises;
const inputBuffer = await fsPromises.readFile("../input.txt");
const input = inputBuffer.toString('utf8');  

let left = [];
let right = [];

const regex = /^(?<left>\d+)\s+(?<right>\d+)$/;

// TODO refactor into something nicer
input.split("\n").reduce((acc, line) =>
	{
		if (line != "") {
			const result = line.match(regex);
			const left_in_number = parseInt(result.groups["left"]);
			const right_in_number = parseInt(result.groups["right"]);
			left.push(left_in_number);
			right.push(right_in_number);
		}
		acc
		
	}
	, 0);

left.sort();
right.sort();

let totalDistance = 0;

left.map((element, index) => [element, right[index]]).reduce((acc, [left, right]) => {
	const diff = Math.abs(left - right);
	totalDistance = totalDistance + diff;
	acc
}, 0);

console.log(totalDistance);
